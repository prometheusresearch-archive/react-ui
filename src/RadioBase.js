/**
 * @copyright 2016-present, Prometheus Research, LLC
 * @flow
 */

import React from 'react';
import {uniqueId, noop} from './Utils';
import * as I18N from './I18N';
import * as Focus from './Focus';

type Props = {
  value: boolean,
  onChange: (boolean) => *,
  label?: string,
  title?: string,
  hint?: string,
  idx: number,
  disabled?: boolean,
  focusIndex?: number,
  inputRef: (HTMLElement) => *,
  variant?: Object,
};

type Component = string | ReactClass<*>;

export type Stylesheet = {
  Root: Component,
  Input: Component,
  LabelWrapper: Component,
  Label: Component,
  Hint: Component,
};

export let stylesheet: Stylesheet = {
  Root: 'div',
  Input: 'input',
  LabelWrapper: 'div',
  Hint: 'div',
  Label: 'div',
};

export default class RadioBase extends React.Component<*, Props, *> {
  ariaId: string;

  static stylesheet = stylesheet;

  static contextTypes = I18N.contextTypes;

  static defaultProps = {
    onChange: noop,
  };

  constructor(props: Props) {
    super(props);
    this.ariaId = uniqueId('aria');
  }

  render() {
    let {
      value,
      label,
      title,
      hint,
      inputRef,
      variant,
      disabled,
      focusIndex,
      idx: _idx,
      ...props
    } = this.props;
    let {
      Root,
      Input,
      Label,
      Hint,
      LabelWrapper,
    } = this.constructor.stylesheet;
    let {i18n = I18N.defaultContext} = this.context;
    variant = {
      rtl: i18n.dir === 'rtl',
      ltr: i18n.dir === 'ltr',
      disabled,
      ...variant,
    };
    return (
      <Root title={title} variant={variant}>
        <Focus.Focusable focusIndex={focusIndex}>
          <Input
            {...props}
            aria-labelledby={this.ariaId}
            disabled={disabled}
            variant={variant}
            ref={inputRef}
            inputRef={inputRef}
            type="radio"
            checked={value}
            onChange={this.onChange}
          />
        </Focus.Focusable>
        {label &&
          <LabelWrapper variant={variant} onClick={this.onClick}>
            <Label id={this.ariaId} variant={variant}>{label}</Label>
            <Hint variant={variant}>{hint}</Hint>
          </LabelWrapper>}
      </Root>
    );
  }

  onClick = () => {
    if (!this.props.disabled) {
      this.props.onChange(true);
    }
  };

  onChange = (e: UIEvent) => {
    // $FlowIssue: ...
    let checked = e.target.checked;
    this.props.onChange(checked);
  };
}
