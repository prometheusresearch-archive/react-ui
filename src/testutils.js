/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';
import {isEqual} from 'lodash/lang';
import indent from 'indent-string';

import {
  findWithType as _findWithType,
  findAllWithType as _findAllWithType,
  getMountedInstance,
  findAll
} from 'react-shallow-testutils';

import {
  createRenderer as _createRenderer
} from 'react-addons-test-utils';

import _inspectElement, {
  inspectReactProp as _inspectReactProp
} from 'inspect-react-element';


import assert from 'power-assert';
import {stub, spy, mock} from 'sinon';

export {stub, spy, mock, assert};

class FormDataMock {

  append = spy();
}

let originalFormData = global.FormData;

export function mockFormData() {
  global.FormData = FormDataMock;
}

export function unmockFormData() {
  global.FormData = originalFormData;
}

function inspectReactProps(node) {
  return '__originalConfig' in node.props ?
    {...node.props.__originalConfig, children: node.props.children} :
    node.props || {};
}

function inspectReactProp(propName, propValue) {
  if (propName === '__originalConfig' || propName === 'key' || propName === 'ref') {
    return null;
  }
  return _inspectReactProp(propName, propValue);
}

export function inspectElement(node) {
  return _inspectElement(node, {inspectReactProps, inspectReactProp});
}

class ReactShallowRenderer {

  constructor() {
    this._renderer = _createRenderer();
  }

  render(element, context) {
    return this._renderer.render(element, context);
  }

  get instance() {
    return getMountedInstance(this._renderer);
  }

  get element() {
    return this._renderer.getRenderOutput();
  }

  findWithType(type) {
    return findWithType(this.element, type);
  }

  findAllWithType(type) {
    return findAllWithType(this.element, type);
  }

  assertElementWithType(type) {
    this.findWithType(type);
  }

  assertNoElementWithType(type) {
    assert(this.findAllWithType(type).length === 0);
  }


  findWithTypeProps(type, props) {
    return findWithTypeProps(this.element, type, props);
  }

  findAllWithTypeProps(type, props) {
    return findAllWithTypeProps(this.element, type, props);
  }

  assertElementWithTypeProps(type, props) {
    this.findWithTypeProps(type, props);
  }

  assertNoElementWithTypeProps(type, props) {
    assert(this.findAllWithTypeProps(type, props).length === 0);
  }


  findWithElement(element) {
    return findWithElement(this.element, element);
  }

  findAllWithElement(element) {
    return findAllWithElement(this.element, element);
  }

  assertElement(element) {
    this.findWithElement(element);
  }

  assertNoElement(element) {
    assert(this.findAllWithElement(element).length === 0);
  }

}

export function createRenderer() {
  return new ReactShallowRenderer();
}

export function findWithType(root, type) {
  let elements = findAll(root, _makeMatchWithType(type));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one partial match for element:
${indent(inspectElement(React.createElement(type)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithType(root, type) {
  return findAll(root, _makeMatchWithType(type));
}

function _makeMatchWithType(type) {
  return function _matchWithType(element) {
    return element && element.type && (
      element.type === type ||
      element.type.Component === type
    );
  };
}

export function findWithTypeProps(root, type, props) {
  let elements = findAll(root, _makeMatchWithElement(type, props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(React.createElement(type, props)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithTypeProps(root, type, props) {
  return findAll(root, _makeMatchWithElement(type, props));
}

export function findWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  let elements = findAll(root, _makeMatchWithElement(spec.type, spec.props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(spec), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  return findAll(root, _makeMatchWithElement(spec.type, spec.props));
}

function _makeMatchWithElement(type, props) {
  let matchWithType = _makeMatchWithType(type);

  let matchWithProps = function(element) {
    for (let key in props) {
      if (!props.hasOwnProperty(key)) {
        continue;
      }
      if (!element.props.hasOwnProperty(key)) {
        continue;
      }
      if (key === '__originalConfig') {
        continue;
      }
      if (!isEqual(element.props[key], props[key])) {
        return false;
      }
    }
    return true;
  };
  return function _matchWithElement(element) {
    return element && matchWithType(element) && matchWithProps(element);
  };
}

function _normalizeReactElementSpec(spec) {
  let {props: {__originalConfig = null}} = spec;
  if (__originalConfig) {
    return {...spec, props: __originalConfig};
  } else {
    return spec;
  }
}

let _createElement = React.createElement;

function createElement(type, config, ...children) {
  config = {...config, __originalConfig: config || {}};
  let element = _createElement(type, config, ...children);
  return element;
}

React.createElement = createElement;

