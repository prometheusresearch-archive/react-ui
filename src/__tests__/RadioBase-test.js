/**
 * @copyright 2016, Prometheus Research, LLC
 */

import React from 'react';

import {createRenderer, spy, assert} from '../testutils';
import RadioBase from '../RadioBase';

describe('react-ui', function() {

  let renderer = null;

  beforeEach(function() {
    renderer = createRenderer();
  });

  describe('<RadioBase />', function() {

    it('triggers onChange on input', function() {
      let onChange = spy();
      renderer.render(
        <RadioBase
          value={false}
          onChange={onChange}
          />
      );

      let input = renderer.findWithType('input');
      assert(input.props.onChange);
      input.props.onChange({target: {checked: true}});
      assert(onChange.callCount === 1);
      assert.deepEqual(onChange.lastCall.args, [true]);
    });

    it('triggers onChange on label click', function() {
      let onChange = spy();
      renderer.render(
        <RadioBase
          label="OK"
          value={false}
          onChange={onChange}
          />
      );

      let divs = renderer.findAllWithType('div');
      let label = divs.find(div => div.props.onClick);
      label.props.onClick();
      assert(onChange.callCount === 1);
      assert.deepEqual(onChange.lastCall.args, [true]);
    });

    it('does not triggerd onChange on label click if disabled', function() {
      let onChange = spy();
      renderer.render(
        <RadioBase
          disabled
          label="OK"
          value={false}
          onChange={onChange}
          />
      );

      let divs = renderer.findAllWithType('div');
      let label = divs.find(div => div.props.onClick);
      label.props.onClick();
      assert(onChange.callCount === 0);
    });
  });

});
