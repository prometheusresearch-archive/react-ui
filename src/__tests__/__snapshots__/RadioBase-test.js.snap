// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not triggerd onChange on label click if disabled 1`] = `
<div
  title={undefined}
  variant={
    Object {
      "disabled": true,
      "ltr": true,
      "rtl": false,
    }
  }
>
  <input
    aria-labelledby="aria3"
    checked={false}
    disabled={true}
    inputRef={[Function]}
    onChange={[Function]}
    type="radio"
    variant={
      Object {
        "disabled": true,
        "ltr": true,
        "rtl": false,
      }
    }
  />
  <div
    onClick={[Function]}
    variant={
      Object {
        "disabled": true,
        "ltr": true,
        "rtl": false,
      }
    }
  >
    <div
      id="aria3"
      variant={
        Object {
          "disabled": true,
          "ltr": true,
          "rtl": false,
        }
      }
    >
      OK
    </div>
    <div
      variant={
        Object {
          "disabled": true,
          "ltr": true,
          "rtl": false,
        }
      }
    />
  </div>
</div>
`;

exports[`triggers onChange on input 1`] = `
<div
  title={undefined}
  variant={
    Object {
      "disabled": undefined,
      "ltr": true,
      "rtl": false,
    }
  }
>
  <input
    aria-labelledby="aria1"
    checked={false}
    disabled={undefined}
    inputRef={[Function]}
    onChange={[Function]}
    type="radio"
    variant={
      Object {
        "disabled": undefined,
        "ltr": true,
        "rtl": false,
      }
    }
  />
</div>
`;

exports[`triggers onChange on label click 1`] = `
<div
  title={undefined}
  variant={
    Object {
      "disabled": undefined,
      "ltr": true,
      "rtl": false,
    }
  }
>
  <input
    aria-labelledby="aria2"
    checked={false}
    disabled={undefined}
    inputRef={[Function]}
    onChange={[Function]}
    type="radio"
    variant={
      Object {
        "disabled": undefined,
        "ltr": true,
        "rtl": false,
      }
    }
  />
  <div
    onClick={[Function]}
    variant={
      Object {
        "disabled": undefined,
        "ltr": true,
        "rtl": false,
      }
    }
  >
    <div
      id="aria2"
      variant={
        Object {
          "disabled": undefined,
          "ltr": true,
          "rtl": false,
        }
      }
    >
      OK
    </div>
    <div
      variant={
        Object {
          "disabled": undefined,
          "ltr": true,
          "rtl": false,
        }
      }
    />
  </div>
</div>
`;
